src/services/projects.js

```diff
@@ -64,10 +86,42 @@ angular.module('openshiftCommonServices')
                               });
                     });
           },
+
+          // List the projects the user has access to. This method returns
+          // cached data if the projects had previously been fetched to avoid
+          // requesting them again and again, which is a problem for admins who
+          // might have hundreds or more.
+          list: function(forceRefresh) {
+            if (cachedProjectData && !forceRefresh) {
+              Logger.debug("ProjectsService: returning cached project data");
+              return $q.when(cachedProjectData);
+            }
+
+            Logger.debug("ProjectsService: listing projects, force refresh", forceRefresh);
+            return DataService.list('projects', {}).then(function(projectData) {
+              cachedProjectData = projectData;
+              return projectData;
```

Comment:
I think it's reasonable for changes outside of the web console to make the user refresh to get the latest list. We could instead re-list if the projects are fewer than a certain cutoff. But for extremely large lists or lists that fail, I don't see a good alternative. Trying to relist every page causes the console to become unusable.

For changes that happen in the web console, we can update the cached list directly. Let me make sure I'm doing that everywhere we can.

I'm looking at all the places we list or watch projects now, and I'm updating them to use the ProjectsService methods.

In 3.8, we'll be able to limit how many results we get back from the server, but not yet in 3.7.